### PR TITLE
Add ast ident iterator

### PR DESCRIPTION
**Pull request by Ryan1729**
While working on #6086 which is a PR meant to address #6039, I found that I needed a way to iterate through the identifiers in an expression. Implementing an `Ident` iterator has turned out to be sufficiently complicated that I thought it would be nicer for reviewers if those changes were in a separate PR. This is that PR.

changelog: none

### CODE REVIEW

**Comment by phansch** [comment]
Thanks for the PR! Unfortunately, I'm not too familiar with this part of the code, so maybe r? @flip1995

**Review by flip1995 — COMMENTED** [review]
So before reviewing all of this: IIUC the goal of this is to get an iterator over all `ident`s in an expression. Is there a reason, that you need lazy evaluation here? Otherwise we could just use a `Visitor`, which collects every `Ident` in a `Vec` and then just turns this `Vec` into an iterator.

Downside is that you have to collect **all** `Ident`s and have a worse performance if you often want to stop iterating after a few steps.

The advantage is, that you don't have to implement all of the expr/item/ty/... handling yourself, since it is already implemented by the `Visitor`.

**Comment by Ryan1729** [comment]
> So before reviewing all of this: IIUC the goal of this is to get an iterator over all `ident`s in an expression. 

That is correct. And the main reason we want an is to create suggestions containing an expression with some of the `Ident`s swapped. Specifically [this code](https://github.com/rust-lang/rust-clippy/blob/c671153f2588858b54bb217f46976fb08018919e/clippy_lints/src/suspicious_operation_groupings.rs#L244-L277) in #6086, (currently a WIP) uses such an iterator: 

> Is there a reason, that you need lazy evaluation here? Otherwise we could just use a `Visitor`, which collects every `Ident` in a `Vec` and then just turns this `Vec` into an iterator.
> 
> Downside is that you have to collect **all** `Ident`s and have a worse performance if you often want to stop iterating after a few steps.
> 
> The advantage is, that you don't have to implement all of the expr/item/ty/... handling yourself, since it is already implemented by the `Visitor`.

Hmm. Well, the part that actually creates the suggestion can stop early. But at least for the purposes of #6086, that iteration would be preceded by some form of full iteration over the expression in question anyway, so maybe that savings would be insignificant. O(n + n / 2) is still O(n).

Given how much work handling all the possibilities myself has been so far, using a `Visitor` sounds like it would be worth doing, given we aren't positive the perf difference. If we only expose an Iterator trait object then if it turns out that laziness is worth doing we can change it without affecting the user(s) of the iterator. So I think I'll given switching to using a `Visitor` and `walk_expr` a go.

File: clippy_lints/src/utils/ast_utils/ident_iter.rs [hunk 1]
```diff
@@ -0,0 +1,1298 @@
...
+            expr,
+            inner: None,
+            done: false,
+        }
+    }
+
+    /// This is a convenience method to help with type inference.
+    fn new_p(expr: &'expr P<Expr>) -> Self {
+        Self::new(expr)
+    }
```
[the remark this answers is not in the document]
Comment (Ryan1729) [reply clippy_lints/src/utils/ast_utils/ident_iter.rs:72]: Given `exprs` is a `&Vec<P<Expr>>`, 

```rust
exprs.iter()
    .flat_map(ExprIdentIter::new_p)
```
compiles, while using `ExprIdentIter::new` does not:
```rust
exprs.iter()
    .flat_map(ExprIdentIter::new)
```

IIRC this compiles:
```rust
exprs.iter()
    .flat_map(|e| ExprIdentIter::new(e))
```
So `ExprIdentIter::new_p` allows skipping writing the closure.
I'm willing to believe there might be a better way to do this. Maybe making `ExprIdentIter::new` generic or something like that?

File: clippy_lints/src/utils/ast_utils/ident_iter.rs [hunk 2]
```diff
@@ -0,0 +1,1298 @@
...
+
+pub type IdentIter<'a> = Box<dyn Iterator<Item = Ident> + 'a>;
+
+pub fn from_expr<'expr>(expr: &'expr Expr) -> IdentIter<'expr> {
+    Box::new(ExprIdentIter::new(expr))
+}
+
+pub fn from_ty<'ty>(ty: &'ty Ty) -> IdentIter<'ty> {
+    Box::new(TyIdentIter::new(ty))
+}
```
[the remark this answers is not in the document]
Comment (Ryan1729) [reply clippy_lints/src/utils/ast_utils/ident_iter.rs:52]: Sounds like a good idea. I'll give that a go.

File: clippy_lints/src/utils/ast_utils/ident_iter.rs [hunk 3]
```diff
@@ -0,0 +1,1298 @@
+use core::iter::{self, FusedIterator};
+use rustc_ast::{
```
[the remark this answers is not in the document]
Comment (Ryan1729) [reply clippy_lints/src/utils/ast_utils/ident_iter.rs]: Done.

File: clippy_lints/src/utils/ast_utils/ident_iter.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ryan1729) [reply clippy_lints/src/utils/ast_utils/ident_iter.rs]: I had to add a struct, because of the orphan rule, but that was easy enough to do.

**Comment by Ryan1729** [comment]
Okay, I feel silly now for not using `Visitor` from the start, given how simple the code has now become.

File: clippy_lints/src/utils/ast_utils/ident_iter.rs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+const ESTIMATED_BYTES_OF_CODE_PER_IDENT: usize = 16;
+
+struct IdentCollector(Vec<Ident>);
+
+impl IdentCollector {
+    fn new(span: Span) -> Self {
+        let byte_count = (span.hi() - span.lo()).0 as usize;
+
+        // bytes / (bytes / idents) = idents
+        IdentCollector(Vec::with_capacity(byte_count / ESTIMATED_BYTES_OF_CODE_PER_IDENT))
```
Comment (flip1995) [inline clippy_lints/src/utils/ast_utils/ident_iter.rs:50]: I really don't think we need to pre-allocate here. Also I don't think we can make a reasonable approximation here.

Without any measurement, we won't know if this even helps. So I would go with the KISS principle over a questionable performance optimization.

**Review by flip1995 — APPROVED** [review]
Since the code is now that simple, I think you can squash it into one commit and include it in the other PR?

File: clippy_lints/src/utils/ast_utils/ident_iter.rs (hunk 4 above)
Comment (Ryan1729) [reply clippy_lints/src/utils/ast_utils/ident_iter.rs]: Fair enough.

**Comment by Ryan1729** [comment]
> Since the code is now that simple, I think you can squash it into one commit and include it in the other PR?

I'll do that. Thanks for all your help simplifying the code!

**Pull request closed without merging**